### ISSUE TITLE
Is there a way to decorate things that are already bound?

### ISSUE THREAD

**Issue by kruncher** [ssannandeji/Zenject-2019#194]
Take the following for instance:
```csharp
this.Container.Bind<IFoo>()
    .To<FooA>()
    .AsWhatever();
```
Elsewhere, probably in a different installer...
```csharp
this.Container.Decorate<IFoo>()
    .With<FooDecoratorA>();
this.Container.Decorate<IFoo>()
    .With<FooDecoratorB>();
```
or perhaps, something like:
```csharp
this.Container.Rebind<IFoo>()
    .With<FooDecoratorA>()
    .AsSingle()
    .WithArguments(this.Container.??????<IFoo>()); // << don't want to resolve here.
this.Container.Rebind<IFoo>()
    .With<FooDecoratorB>()
    .AsSingle()
    .WithArguments(this.Container.??????<IFoo>()); // << don't want to resolve here.
```
Where types are something like:
```csharp
public interface IFoo {
    void Bar();
}

public sealed class FooA : IFoo {
    public void Bar() { ... }
}

public sealed class FooDecoratorA : IFoo {
    private readonly IFoo inner;
    private readonly IBaz baz;

    public FooDecoratorA(IFoo inner, IBaz baz) {
        this.inner = inner;
        this.baz = baz;
    }

    public void Bar() {
        thiz.baz.Qux();
        this.inner.Bar();
    }
}

public sealed class FooDecoratorB : IFoo {
    private readonly IFoo inner;

    public FooDecoratorB(IFoo inner) {
        this.inner = inner;
    }

    public void Bar() {
        Debug.Log("Barring...");
        this.inner.Bar();
        Debug.Log("Barred!");
    }
}
```

**Comment by kruncher** [comment]
To provide some context this allows me to extend functionality in latter installers by decorating whatever dependencies have already been bound.

TL;DR; This reply is basically just showing how I'm using Zenject and installers. If this is of interest please read on; otherwise... avoid! :D

Here is what my service configuration assets look like; you may notice that accolades (achievements and leaderboards) and ads have "Null" installers (null object service installer) since 'standalone' desktop builds do not have these things:

![image](https://cloud.githubusercontent.com/assets/554215/22255889/cf76281c-e250-11e6-9f95-079469881c46.png)

When multiple installers are available for a service a drop-down menu is presented allowing the developer to specify the service installer that they wish to use:

![image](https://cloud.githubusercontent.com/assets/554215/22256910/137e4366-e254-11e6-946d-43e34e867264.png)

A module can provide multiple installer options for it's services, or another module can provide alternative installers for the services defined by other modules.

Modules are structured as follows (directory structure):
```
Assets/
|-- _Modules/
    |-- ModuleA/
        |-- Services/
        |   |-- MyServiceA.cs
        |   |-- MyServiceB.cs
        |-- Foo/
        |   |-- *.cs
        |-- Installers/
        |   |-- MyServiceA_Null_Installer.cs  // 'null' implementation of MyServiceA
        |-- ModuleA.cs
```

Currently I am using [git subtree](https://blogs.atlassian.com/2013/05/alternatives-to-git-submodule-git-subtree/) to import my modules; however in the future I may switch to [npm](https://github.com/npm/npm) or use a Unity-specific package manager should one ever arise. Each module currently has an empty class that is only used to document any other modules that the module is dependent upon. I would not be using these if using something like npm since `package.json` deals with this aspect.
```csharp
[Dependency(typeof(SomeOtherModule))]
public sealed class ModuleA : Module
{
}
```

Services are defined as follows; again these are empty classes, but these essentially declare unique handles for services and are also used to resolve the order of service installers:
```csharp
public sealed class MyServiceA : Service
{
}

[Dependency(typeof(MyServiceA))]
public sealed class MyServiceB : Service
{
}
```

Installers are pretty much the same as normal; except they declare which service they are designed to install. These are `ScriptableObject` type Zenject installers:
```csharp
public class MyServiceA_Null_Installer : ZenjectServiceInstaller<MyServiceA>
{
    public override void InstallBindings()
    {
        this.Container.Bind<IFoo>()
            .To<Foo_Null>()
            .AsSingle();
    }
}
```

Installers optionally can declare additional dependencies on other services as well:
```csharp
[Dependency(typeof(SomeOtherService))]
public class MyServiceB_Bar_Installer : ZenjectServiceInstaller<MyServiceB>
{
    public override void InstallBindings()
    {
        this.Container.Bind<IBarManager>()
            .To<BarManager>()
            .AsSingle();
        this.Container.Bind<IBar>()
            .To<Bar>()
            .AsSingle();
    }
}
```

Since installers are sorted so that dependency services are installed first it would be useful to be able to have later installers that rebind existing services by decorating the previously bound ones:

Service configurations can also inherit other service configurations; this is useful for situations where you want to use different service installers or configurations for different target platforms. At the moment configurations can only inherit installers; they cannot override them:
```
Assets/
|-- _Project/
|   |-- Configurations/
|   |   |-- Targets/
|   |   |   |-- _TargetConfigurationSelector_.asset (indicates current target, set by uTomate or custom build menu)
|   |   |   |-- AmazonKindleConfiguration.asset
|   |   |   |-- AppStoreConfiguration.asset
|   |   |   |-- GooglePlayStoreConfiguration.asset
|   |   |   |-- StandaloneConfiguration.asset
|   |   |-- CommonConfiguration.asset
```
In this case all 4 target configurations inherit any services installed by the common configuration.

With a few minor tweaks my service configuration assets could easily override the installers of inherited configurations. Either to customize values or to provide an alternative installer for a specific build target. In order to do this I just need to automate the construction of a "flattened" configuration asset so that Unity doesn't hoist unwanted assets into builds (i.e. textures, models, prefabs, etc that are referenced by overridden installers).

**Comment by Jamjardavies** [comment]
I'm not sure how to help here, but I did want to say there is a Unity library-like manager that Steve has also done called Projeny.

On a side note: Really like the installer configuration thing you have, is this going to be open-sourced?

**Comment by kruncher** [comment]
@Jamjardavies Yes, depending upon how well it works out in practice, I plan to open source the service installer / configuration assets. I am trialing this in a real project at the moment to iron out kinks.

If there isn't a way to decorate existing bindings by adding decoration logic to a new installer, then that might be a pretty big roadblock. I'm not entirely sure how I would workaround this.

**Comment by svermeulen** [comment]
Here's one idea that may or may not work for you:

Add the following to the DiContainer class:

```
        public ConcreteBinderGeneric<TContract> DecorateBind<TContract>()
        {
            return DecorateBind<TContract>(null);
        }

        public ConcreteBinderGeneric<TContract> DecorateBind<TContract>(object identifier)
        {
            FlushBindings();

            var bindingId = new BindingId(typeof(TContract), identifier);

            var oldProviders = _providers[bindingId];
            _providers.RemoveWithConfirm(bindingId);

            var subContainer = CreateSubContainer();
            subContainer._providers.Add(bindingId, oldProviders);

            var subId = Guid.NewGuid();

            Bind<TContract>().WithId(identifier).FromMethod(
                (ctx) => subContainer.Resolve<TContract>(subId));

            return subContainer.Bind<TContract>().WithId(subId);
        }
```

Then you can use it like this

```
    public interface IFoo
    {
        void Run();
    }

    public class FooA : IFoo
    {
        public void Run()
        {
            Log.Trace("FooA.Run");
        }
    }

    public class FooDecorator : IFoo
    {
        IFoo _foo;

        public FooDecorator(IFoo foo)
        {
            _foo = foo;
        }

        public void Run()
        {
            Log.Trace("FooDecorator.Run (before)");
            _foo.Run();
            Log.Trace("FooDecorator.Run (after)");
        }
    }

    public class Bar : IInitializable
    {
        readonly IFoo _foo;

        public Bar(IFoo foo)
        {
            _foo = foo;
        }

        public void Initialize()
        {
            _foo.Run();
        }
    }

    public class TestInstaller : MonoInstaller<TestInstaller>
    {
        public override void InstallBindings()
        {
            Container.Bind<IFoo>().To<FooA>().AsCached();

            Container.DecorateBind<IFoo>().To<FooDecorator>().AsCached();

            Container.BindInterfacesAndSelfTo<Bar>().AsSingle();
        }
    }
```

Does that get you what you need?

**Comment by kruncher** [comment]
@svermeulen thanks for your reply!

I've done a few basic initial experiments using the methods that you supplied and it seems to be working perfectly!

Would this support the `.WhenXYZ` conditional injection as well?

Many thanks!!!

**Comment by svermeulen** [comment]
Yes, When conditions should work fine.  I'll try and put DecorateBind in the next version.

**Comment by svermeulen** [comment]
Assuming you are talking about adding the .When conditional on to the DecoratorBind then yes that should work.  If you add a When Conditional to the original bind though that wouldn't work with this approach.

**Comment by kruncher** [comment]
Thanks for adding `DecorateBind` to v5.0!

> Assuming you are talking about adding the .When conditional on to the DecoratorBind then yes that should work. If you add a When Conditional to the original bind though that wouldn't work with this approach.

I was referring to the original bind. For instance, allowing a specific prior binding to be decorated.

For me personally this is not too much of an issue. But if using modular installers then it could be  easy to inadvertently remove other bindings that were previously conditionally bound. I see 2 potential use cases:

Use Case 1 - Decorate an existing binding, but only where a specific criteria is met:
```csharp
// Some other binding that we don't want to touch:
this.Container.Bind<IFoo>()
    .To<Foo>()
    .AsSingle()
    .WhenInjectedInto<Bar>();

// The target binding:
this.Container.Bind<IFoo>()
    .To<Foo2>()
    .AsTransient()
    .WhenInjectedInto<Qux>();

// One 'FooDecorator' instance created for each previous Foo2 instance.
this.Container.DecorateBind<IFoo>()
    .With<FooDecorator>()
    .WithArguments(someArguments)
    .WhenInjectedInto<Qux>();   // .AsXyz inherited from original binding
```

Use Case 2 - Decorate all existing bindings (regardless of any specific criteria):
```csharp
this.Container.Bind<IFoo>()
    .To<Foo>()
    .AsSingle()
    .WhenInjectedInto<Bar>();

this.Container.Bind<IFoo>()
    .To<Foo2>()
    .AsTransient()
    .WhenInjectedInto<Qux>();

// One 'FooDecorator' instance created for each previous Foo and Foo2 instance.
this.Container.DecorateBind<IFoo>()
    .With<FooDecorator>()   // .AsXyz inherited from original binding
    .WithArguments(someArguments);
```

In the above example `.With` is basically the same as `.To`; except it doesn't allow the `AsXyz` methods to be specified since decorator instances should always be 1:1 to the original resolved bindings.

**Comment by svermeulen** [comment]
Ok I've started looking at cleaning this up to the point where it can be merged

I think you summed up the issues nicely with your two use cases.  I'm not sure we could cover case 1 though.  We might have to just disallow specifying a When condition with DecoratorBind and then always assume the decorator applies to all IFoo bindings.  Since, if we wanted to support case 1, we'd have to compare the When() conditionals between the decorator bind and all the normal binds, so we know where to apply the decorator, which might be difficult/error prone/not possible.

At the very least though, I think case 2 is doable.  I agree that DecorateBind should just use the same AsXyz scope for each binding that it decorates like you say

**Comment by whatwapp** [comment]
Hi,
any update on this? I am interested too in this function.

I'd prefer to not modify DiContainer directly, is there a way to do this with an extension method until this feature is implemented properly?

**Comment by svermeulen** [comment]
This has been added now to zenject 6.  Currently undocumented, but it looks like this:

    Container.Bind<ISaveHandler>().To<SaveHandler>().AsSingle();

    Container.Decorate<ISaveHandler>().With<SaveDecorator1>();
    Container.Decorate<ISaveHandler>().With<SaveDecorator2>().FromX().WithArguments(..);

**Comment by kruncher** [comment]
Awesome!! thanks :D

I am so pleased that Zenject still relieves lots of love; my go to asset for all Unity projects now!

**Comment by svermeulen** [comment]
Glad you like it  :)   Yeah, it sometimes takes me awhile to find time to address these bugs but I always return to it eventually
